File: BusinessLogic/ItemType.kt
package com.hamthelegend.enchantmentorder.android.domain.model.item

import androidx.annotation.DrawableRes
import com.hamthelegend.enchantmentorder.android.R
import com.hamthelegend.enchantmentorder.android.domain.calculator.plus
import com.hamthelegend.enchantmentorder.android.domain.model.enchantment.EnchantmentType
import com.hamthelegend.enchantmentorder.android.domain.model.enchantment.EnchantmentType.*

enum class ItemType(
    val friendlyName: String,
    val compatibleEnchantmentTypes: Set<EnchantmentType>,
    val defaultEnchantmentTypes: Set<EnchantmentType>,
    val itemCategory: ItemCategory,
    @DrawableRes val iconResource: Int,
) {
    EnchantedBook(
        friendlyName = "Enchanted Book",
        compatibleEnchantmentTypes = EnchantmentType.values().toSet(),
        defaultEnchantmentTypes = setOf(),
        itemCategory = ItemCategory.ENCHANTED_BOOK,
        iconResource = R.drawable.book_enchanted,
    ),
    Helmet(
        friendlyName = "Helmet",
        compatibleEnchantmentTypes = setOf(
            AquaAffinity,
            BlastProtection,
            CurseOfBinding,
            CurseOfVanishing,
            FireProtection,
            Mending,
            ProjectileProtection,
            Protection,
            Respiration,
            Thorns,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            AquaAffinity,
            Mending,
            Protection,
            Respiration,
            Thorns,
            Unbreaking,
        ),
        itemCategory = ItemCategory.WEARABLE,
        iconResource = R.drawable.helmet,
    ),
    Chestplate(
        friendlyName = "Chestplate",
        compatibleEnchantmentTypes = setOf(
            BlastProtection,
            CurseOfBinding,
            CurseOfVanishing,
            FireProtection,
            Mending,
            ProjectileProtection,
            Protection,
            Thorns,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            Mending,
            Protection,
            Thorns,
            Unbreaking,
        ),
        itemCategory = ItemCategory.WEARABLE,
        iconResource = R.drawable.chestplate,
    ),
    Leggings(
        friendlyName = "Leggings",
        compatibleEnchantmentTypes = setOf(
            BlastProtection,
            CurseOfBinding,
            CurseOfVanishing,
            FireProtection,
            Mending,
            ProjectileProtection,
            Protection,
            Thorns,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            Mending,
            Protection,
            SwiftSneak,
            Thorns,
            Unbreaking,
        ),
        itemCategory = ItemCategory.WEARABLE,
        iconResource = R.drawable.leggings,
    ),
    Boots(
        friendlyName = "Boots",
        compatibleEnchantmentTypes = setOf(
            BlastProtection,
            CurseOfBinding,
            CurseOfVanishing,
            DepthStrider,
            FeatherFalling,
            FireProtection,
            FrostWalker,
            Mending,
            ProjectileProtection,
            Protection,
            SoulSpeed,
            Thorns,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            DepthStrider,
            FeatherFalling,
            Mending,
            Protection,
            SoulSpeed,
            Thorns,
            Unbreaking,
        ),
        itemCategory = ItemCategory.WEARABLE,
        iconResource = R.drawable.boots,
    ),
    Elytra(
        friendlyName = "Elytra",
        compatibleEnchantmentTypes = setOf(
            CurseOfBinding,
            CurseOfVanishing,
            Mending,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            Mending,
            Unbreaking,
        ),
        itemCategory = ItemCategory.WEARABLE,
        iconResource = R.drawable.elytra,
    ),
    Head(
        friendlyName = "Head",
        compatibleEnchantmentTypes = setOf(
            CurseOfBinding,
            CurseOfVanishing,
        ),
        defaultEnchantmentTypes = setOf(
            CurseOfBinding,
        ),
        itemCategory = ItemCategory.WEARABLE,
        iconResource = R.drawable.head,
    ),
    Sword(
        friendlyName = "Sword",
        compatibleEnchantmentTypes = setOf(
            BaneOfArthropods,
            CurseOfVanishing,
            FireAspect,
            Knockback,
            Looting,
            Mending,
            Sharpness,
            Smite,
            SweepingEdge,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            FireAspect,
            Knockback,
            Looting,
            Mending,
            Sharpness,
            SweepingEdge,
            Unbreaking,
        ),
        itemCategory = ItemCategory.TOOL,
        iconResource = R.drawable.sword,
    ),
    Axe(
        friendlyName = "Axe",
        compatibleEnchantmentTypes = setOf(
            BaneOfArthropods,
            CurseOfVanishing,
            Efficiency,
            Fortune,
            Mending,
            Sharpness,
            SilkTouch,
            Smite,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            Efficiency,
            Mending,
            Sharpness,
            SilkTouch,
            Unbreaking,
        ),
        itemCategory = ItemCategory.TOOL,
        iconResource = R.drawable.axe,
    ),
    Pickaxe(
        friendlyName = "Pickaxe",
        compatibleEnchantmentTypes = setOf(
            CurseOfVanishing,
            Efficiency,
            Fortune,
            Mending,
            SilkTouch,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            Efficiency,
            Mending,
            SilkTouch,
            Unbreaking,
        ),
        itemCategory = ItemCategory.TOOL,
        iconResource = R.drawable.pickaxe,
    ),
    Shovel(
        friendlyName = "Shovel",
        compatibleEnchantmentTypes = setOf(
            CurseOfVanishing,
            Efficiency,
            Fortune,
            Mending,
            SilkTouch,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            Efficiency,
            Mending,
            SilkTouch,
            Unbreaking,
        ),
        itemCategory = ItemCategory.TOOL,
        iconResource = R.drawable.shovel,
    ),
    Hoe(
        friendlyName = "Hoe",
        compatibleEnchantmentTypes = setOf(
            CurseOfVanishing,
            Efficiency,
            Fortune,
            Mending,
            SilkTouch,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            Efficiency,
            Mending,
            SilkTouch,
            Unbreaking,
        ),
        itemCategory = ItemCategory.TOOL,
        iconResource = R.drawable.hoe,
    ),
    Bow(
        friendlyName = "Bow",
        compatibleEnchantmentTypes = setOf(
            CurseOfVanishing,
            Flame,
            Infinity,
            Mending,
            Power,
            Punch,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            Flame,
            Infinity,
            Power,
            Punch,
            Unbreaking,
        ),
        itemCategory = ItemCategory.TOOL,
        iconResource = R.drawable.bow,
    ),
    FishingRod(
        friendlyName = "Fishing Rod",
        compatibleEnchantmentTypes = setOf(
            CurseOfVanishing,
            LuckOfTheSea,
            Lure,
            Mending,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            LuckOfTheSea,
            Lure,
            Mending,
            Unbreaking,
        ),
        itemCategory = ItemCategory.TOOL,
        iconResource = R.drawable.fishing_rod,
    ),
    Trident(
        friendlyName = "Trident",
        compatibleEnchantmentTypes = setOf(
            Channeling,
            CurseOfVanishing,
            Impaling,
            Loyalty,
            Mending,
            Riptide,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            Channeling,
            Impaling,
            Loyalty,
            Mending,
            Unbreaking,
        ),
        itemCategory = ItemCategory.TOOL,
        iconResource = R.drawable.trident,
    ),
    Crossbow(
        friendlyName = "Crossbow",
        compatibleEnchantmentTypes = setOf(
            CurseOfVanishing,
            Mending,
            Multishot,
            Piercing,
            QuickCharge,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            Mending,
            Multishot,
            QuickCharge,
            Unbreaking,
        ),
        itemCategory = ItemCategory.TOOL,
        iconResource = R.drawable.crossbow,
    ),
    Shears(
        friendlyName = "Shears",
        compatibleEnchantmentTypes = setOf(
            CurseOfVanishing,
            Efficiency,
            Mending,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            Efficiency,
            Mending,
            Unbreaking,
        ),
        itemCategory = ItemCategory.TOOL,
        iconResource = R.drawable.shears,
    ),
    FlintAndSteel(
        friendlyName = "Flint and Steel",
        compatibleEnchantmentTypes = setOf(
            CurseOfVanishing,
            Mending,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            Mending,
            Unbreaking,
        ),
        itemCategory = ItemCategory.TOOL,
        iconResource = R.drawable.flint_and_steel,
    ),
    CarrotOnAStick(
        friendlyName = "Carrot on a Stick",
        compatibleEnchantmentTypes = setOf(
            CurseOfVanishing,
            Mending,
            Unbreaking,
        ),
        defaultEnchantmentTypes = setOf(
            Mending,
            Unbreaking,
        ),
        itemCategory = ItemCategory.TOOL,
        iconResource = R.drawable.carrot_on_a_stick,
    ),
    WarpedFungusOnAStick(
        friendlyName = "Warped Fungus on a Stick",
        compatibleEnchantmentTypes = setOf(
            CurseOfVanishing,
            Mending,
            Unbreaking
        ),
        defaultEnchantmentTypes = setOf(
            Mending,
            Unbreaking,
        ),
        itemCategory = ItemCategory.TOOL,
        iconResource = R.drawable.warped_fungus_on_a_stick,
    ),
    ;

    infix fun isCompatibleWith(target: ItemType): Boolean {
        val sacrifice = this
        return target == sacrifice || sacrifice == EnchantedBook
    }

    infix fun isIncompatibleWith(target: ItemType) = !isCompatibleWith(target)

    fun supposedProduct(books: List<Item>): Item {
        var item = new(this)
        for (book in books) {
            val product = (item + book).product
            item = product
        }
        return item
    }

}

val targetableItemTypes = ItemType.values().filter {
    (it.itemCategory == ItemCategory.TOOL || it.itemCategory == ItemCategory.WEARABLE)
}

File: BusinessLogic/ItemType.cs
namespace BusinessLogic;

public record ItemType(
    string FriendlyName,
    List<EnchantmentType> CompatibleEnchantmentTypes,
    List<EnchantmentType> DefaultEnchantmentTypes,
    Uri ImageUri)
{
    public static readonly ItemType EnchantedBook = new(
        FriendlyName: "Enchanted Book",
        CompatibleEnchantmentTypes: EnchantmentType.All,
        DefaultEnchantmentTypes: new List<EnchantmentType>(),
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    // TODO: Change the image URIs to match the actual images under the Assets folder that represent the items

    public static readonly ItemType Helmet = new(
       FriendlyName: "Helmet",
       CompatibleEnchantmentTypes: new List<EnchantmentType>
       {
            EnchantmentType.AquaAffinity,
            EnchantmentType.BlastProtection,
            EnchantmentType.CurseOfBinding,
            EnchantmentType.CurseOfVanishing,
            EnchantmentType.FireProtection,
            EnchantmentType.Mending,
            EnchantmentType.ProjectileProtection,
            EnchantmentType.Protection,
            EnchantmentType.Respiration,
            EnchantmentType.Thorns,
            EnchantmentType.Unbreaking
       },
       DefaultEnchantmentTypes: new List<EnchantmentType>
       {
            EnchantmentType.AquaAffinity,
            EnchantmentType.Mending,
            EnchantmentType.Protection,
            EnchantmentType.Respiration,
            EnchantmentType.Thorns,
            EnchantmentType.Unbreaking
       },
       ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType Chestplate = new(
       FriendlyName: "Chestplate",
       CompatibleEnchantmentTypes: new List<EnchantmentType>
       {
            EnchantmentType.BlastProtection,
            EnchantmentType.CurseOfBinding,
            EnchantmentType.CurseOfVanishing,
            EnchantmentType.FireProtection,
            EnchantmentType.Mending,
            EnchantmentType.ProjectileProtection,
            EnchantmentType.Protection,
            EnchantmentType.Thorns,
            EnchantmentType.Unbreaking
       },
       DefaultEnchantmentTypes: new List<EnchantmentType>
       {
            EnchantmentType.Mending,
            EnchantmentType.Protection,
            EnchantmentType.Thorns,
            EnchantmentType.Unbreaking
       },
       ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType Leggings = new(
       FriendlyName: "Leggins",
       CompatibleEnchantmentTypes: new List<EnchantmentType>
       {
                EnchantmentType.BlastProtection,
                EnchantmentType.CurseOfBinding,
                EnchantmentType.CurseOfVanishing,
                EnchantmentType.FireProtection,
                EnchantmentType.Mending,
                EnchantmentType.ProjectileProtection,
                EnchantmentType.Protection,
                EnchantmentType.Thorns,
                EnchantmentType.Unbreaking,
                EnchantmentType.SwiftSneak
       },
       DefaultEnchantmentTypes: new List<EnchantmentType>
       {
                EnchantmentType.Mending,
                EnchantmentType.Protection,
                EnchantmentType.SwiftSneak,
                EnchantmentType.Thorns,
                EnchantmentType.Unbreaking
       },
       ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType Boots = new(
       FriendlyName: "Boots",
       CompatibleEnchantmentTypes: new List<EnchantmentType>
       {
                EnchantmentType.BlastProtection,
                EnchantmentType.CurseOfBinding,
                EnchantmentType.CurseOfVanishing,
                EnchantmentType.DepthStrider,
                EnchantmentType.FeatherFalling,
                EnchantmentType.FireProtection,
                EnchantmentType.FrostWalker,
                EnchantmentType.Mending,
                EnchantmentType.ProjectileProtection,
                EnchantmentType.Protection,
                EnchantmentType.SoulSpeed,
                EnchantmentType.Thorns,
                EnchantmentType.Unbreaking
       },
       DefaultEnchantmentTypes: new List<EnchantmentType>
       {
                EnchantmentType.DepthStrider,
                EnchantmentType.FeatherFalling,
                EnchantmentType.Mending,
                EnchantmentType.Protection,
                EnchantmentType.SoulSpeed,
                EnchantmentType.Thorns,
                EnchantmentType.Unbreaking
       },
       ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType Elytra = new(
       FriendlyName: "Elytra",
       CompatibleEnchantmentTypes: new List<EnchantmentType>
       {
                EnchantmentType.CurseOfBinding,
                EnchantmentType.CurseOfVanishing,
                EnchantmentType.Mending,
                EnchantmentType.Unbreaking
       },
       DefaultEnchantmentTypes: new List<EnchantmentType>
       {
                EnchantmentType.Mending,
                EnchantmentType.Unbreaking
       },
       ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType Head = new(
       FriendlyName: "Head",
       CompatibleEnchantmentTypes: new List<EnchantmentType>
       {
                EnchantmentType.CurseOfBinding,
                EnchantmentType.CurseOfVanishing,

       },
   DefaultEnchantmentTypes: new List<EnchantmentType>
       {
                EnchantmentType.CurseOfBinding,
       },
       ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));
    
    public static readonly ItemType Sword = new(
        FriendlyName: "Sword",
        CompatibleEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.BaneOfArthropods,
            EnchantmentType.CurseOfVanishing,
            EnchantmentType.FireAspect,
            EnchantmentType.Knockback,
            EnchantmentType.Looting,
            EnchantmentType.Mending,
            EnchantmentType.Sharpness,
            EnchantmentType.Smite,
            EnchantmentType.SweepingEdge,
            EnchantmentType.Unbreaking
        },
        DefaultEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.FireAspect,
            EnchantmentType.Knockback,
            EnchantmentType.Looting,
            EnchantmentType.Mending,
            EnchantmentType.Sharpness,
            EnchantmentType.SweepingEdge,
            EnchantmentType.Unbreaking
        },
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType Axe = new(
        FriendlyName: "Axe",
        CompatibleEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.BaneOfArthropods,
            EnchantmentType.CurseOfVanishing,
            EnchantmentType.Efficiency,
            EnchantmentType.Fortune,
            EnchantmentType.Mending,
            EnchantmentType.Sharpness,
            EnchantmentType.SilkTouch,
            EnchantmentType.Smite,
            EnchantmentType.Unbreaking
        },
        DefaultEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.Efficiency,
            EnchantmentType.Mending,
            EnchantmentType.Sharpness,
            EnchantmentType.SilkTouch,
            EnchantmentType.Unbreaking
        },
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType Pickaxe = new(
        FriendlyName: "Pickaxe",
        CompatibleEnchantmentTypes: new List<EnchantmentType>
        {
             EnchantmentType.CurseOfVanishing,
             EnchantmentType.Efficiency,
             EnchantmentType.Fortune,
             EnchantmentType.Mending,
             EnchantmentType.SilkTouch,
             EnchantmentType.Unbreaking
        },
        DefaultEnchantmentTypes: new List<EnchantmentType>
        {
             EnchantmentType.Efficiency,
             EnchantmentType.Mending,
             EnchantmentType.SilkTouch,
             EnchantmentType.Unbreaking
        },
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType Shovel = new(
        FriendlyName: "Shovel",
        CompatibleEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.CurseOfVanishing,
            EnchantmentType.Efficiency,
            EnchantmentType.Fortune,
            EnchantmentType.Mending,
            EnchantmentType.SilkTouch,
            EnchantmentType.Unbreaking
        },
        DefaultEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.Efficiency,
            EnchantmentType.Mending,
            EnchantmentType.SilkTouch,
            EnchantmentType.Unbreaking
        },
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType Hoe = new(
        FriendlyName: "Hoe",
        CompatibleEnchantmentTypes: new List<EnchantmentType>
        {
             EnchantmentType.CurseOfVanishing,
             EnchantmentType.Efficiency,
             EnchantmentType.Fortune,
             EnchantmentType.Mending,
             EnchantmentType.SilkTouch,
             EnchantmentType.Unbreaking
        },
        DefaultEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.Efficiency,
             EnchantmentType.Mending,
             EnchantmentType.SilkTouch,
             EnchantmentType.Unbreaking
        },
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType Bow = new(
        FriendlyName: "Bow",
        CompatibleEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.CurseOfVanishing,
            EnchantmentType.Flame,
            EnchantmentType.Infinity,
            EnchantmentType.Mending,
            EnchantmentType.Power,
            EnchantmentType.Punch,
            EnchantmentType.Unbreaking
        },
        DefaultEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.Flame,
            EnchantmentType.Infinity,
            EnchantmentType.Power,
            EnchantmentType.Punch,
            EnchantmentType.Unbreaking
        },
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType FishingRod = new(
        FriendlyName: "FishingRod",
        CompatibleEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.CurseOfVanishing,
            EnchantmentType.LuckOfTheSea,
            EnchantmentType.Lure,
            EnchantmentType.Mending,
            EnchantmentType.Unbreaking
        },
        DefaultEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.LuckOfTheSea,
            EnchantmentType.Lure,
            EnchantmentType.Mending,
            EnchantmentType.Unbreaking
        },
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType Trident = new(
        FriendlyName: "Trident",
        CompatibleEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.Channeling,
            EnchantmentType.CurseOfVanishing,
            EnchantmentType.Impaling,
            EnchantmentType.Loyalty,
            EnchantmentType.Mending,
            EnchantmentType.Riptide,
            EnchantmentType.Unbreaking
        },
        DefaultEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.Channeling,
            EnchantmentType.Impaling,
            EnchantmentType.Loyalty,
            EnchantmentType.Mending,
            EnchantmentType.Unbreaking
        },
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType Crossbow = new(
        FriendlyName: "Crossbow",
        CompatibleEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.CurseOfVanishing,
            EnchantmentType.Mending,
            EnchantmentType.Multishot,
            EnchantmentType.Piercing,
            EnchantmentType.QuickCharge,
            EnchantmentType.Unbreaking
        },
        DefaultEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.Mending,
            EnchantmentType.Multishot,
            EnchantmentType.QuickCharge,
            EnchantmentType.Unbreaking
        },
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType Shears = new(
        FriendlyName: "Shears",
        CompatibleEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.CurseOfVanishing,
            EnchantmentType.Efficiency,
            EnchantmentType.Mending,
            EnchantmentType.Unbreaking
        },
        DefaultEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.Efficiency,
            EnchantmentType.Mending,
            EnchantmentType.Unbreaking
        },
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType FlintAndSteel = new(
        FriendlyName: "FlintAndSteel",
        CompatibleEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.CurseOfVanishing,
            EnchantmentType.Mending,
            EnchantmentType.Unbreaking
        },
        DefaultEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.Mending,
            EnchantmentType.Unbreaking
        },
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType CarrotOnAStick = new(
        FriendlyName: "CarrotOnAStick",
        CompatibleEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.CurseOfVanishing,
            EnchantmentType.Mending,
            EnchantmentType.Unbreaking
        },
        DefaultEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.Mending,
            EnchantmentType.Unbreaking
        },
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly ItemType WarpedFungusOnAStick = new(
        FriendlyName: "WarpedFungusOnAStick",
        CompatibleEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.CurseOfVanishing,
            EnchantmentType.Mending,
            EnchantmentType.Unbreaking
        },
        DefaultEnchantmentTypes: new List<EnchantmentType>
        {
            EnchantmentType.Mending,
            EnchantmentType.Unbreaking
        },
        ImageUri: new Uri("ms-appx:///Assets/book_enchanted.png"));

    public static readonly List<ItemType> All = new()
    {
        EnchantedBook,
        Helmet,
        Chestplate,
        Leggings,
        Boots,
        Elytra,
        Head,
        Sword,
        Axe,
        Pickaxe,
        Shovel,
        Hoe,
        Bow,
        FishingRod,
        Trident,
        Crossbow,
        Shears,
        FlintAndSteel,
        CarrotOnAStick,
        WarpedFungusOnAStick,
    };
    
    public bool IsCompatibleWith(ItemType target)
    {
        var sacrifice = this;
        return target == sacrifice || sacrifice == EnchantedBook;
    }

    public bool IsIncompatibleWith(ItemType target) => !IsCompatibleWith(target);

    public Item SupposedProduct(List<Item> books)
    {
        var item = this.ToNewItem();
        foreach (var book in books)
        {
            var product = (item + book).Product;
            item = product;
        }

        return item;
    }

    public static List<ItemType> Targetable = ItemType.All.Where(item => item != EnchantedBook).ToList();

}
